code/modules/GOAP/GOAPAgent.dm

```diff
@@ -0,0 +1,241 @@
+GLOBAL_LIST_INIT(goap_smashable_objs, typecacheof(list(
+	/obj/machinery,
+	/obj/structure/window,
+	/obj/structure/closet,
+	/obj/structure/table,
+	/obj/structure/grille,
+	/obj/structure/girder,
+	/obj/structure/rack,
+	/obj/structure/barricade)))
+
+GLOBAL_LIST_INIT(dangerous_turfs, typecacheof(list(
+	/turf/open/floor/plating/lava,
+	/turf/open/chasm)))
+
+
+#define STATE_IDLE		0
+#define STATE_MOVINGTO	1
+#define STATE_ACTING	2
+
+#define GOAP_DEBUG_ENABLE 0
+
+#if GOAP_DEBUG_ENABLE
+	#define goap_debug(X) world.log << "GOAP: [X]"
+#else
+	#define goap_debug(X) ;
+
+/datum/goap_agent
+	var/brain_state = STATE_IDLE
+	var/list/our_actions //The actions available to us (/datum/goap_action)
+	var/list/action_queue //The actions of our current plan (if any) (/datum/goap_action)
+
+	//How we plan (Imagine a specific planner that ALWAYS prefers actions involving murder, regardless of cost, bloodlust agent!)
+	var/datum/goap_planner/planner = /datum/goap_planner
+
+	//Provides Info to the agent
+	var/datum/goap_info_provider/info = /datum/goap_info_provider
+
+	var/atom/agent //The actual atom that uses this brain (who knows, maybe you want to give a donut intelligence?)
+
+	var/list/path
+	var/turf/dest
+	var/turf/last_node
+	var/tries = 0
+	var/obj/item/weapon/card/id/given_pathfind_access
+	var/movement_type = 1 // 1 = normal A*(expensive but near perfect), 2 = Fake A*(rarely works), 3 = Dumb Movement(cheapest), 4 = A* for Lavaland Mobs(use if Enviro-Smash + advanced pathfinding is needed)
+	var/turf/current_loc
+	var/is_megafauna = FALSE
+	var/actions_halted = FALSE
+	var/already_acting = FALSE
+
+	var/fuck_you_astar = FALSE
```

Comment:
I prefer https://github.com/tgstation/tgstation/blob/16c1ad5c289f269250ec28e44e380f423296e9e3/code/modules/atmospherics/machinery/atmosmachinery.dm#L179
'fuck_your_dir', the dir in which you get fucked.